### PR TITLE
fix(ivy): walk declaration views in listener

### PR DESCRIPTION
**Pull request by kara**
This PR moves `nextContext` calls from the top of creation mode to listener functions. This ensures that we're not doing more work than we have to in creation mode (e.g. listener may never fire, save startup time).

In order to move the context instructions into the handler, the previous view must be restored. Since handlers don't fire during change detection, the normal current view won't be available.

There are two new instructions: 
- `getCurrentView`: returns a snapshot of the current view in creation mode
- `restoreView`: restores the given view as `contextViewData`

### CODE REVIEW

File: packages/core/src/render3/instructions.ts [hunk 1]
```diff
@@ -108,11 +108,31 @@ export function getCurrentSanitizer(): Sanitizer|null {
...
 
-export function getViewData(): LViewData {
+/**
+ * Returns the current LViewData instance.
+ *
+ * Used in conjunction with the restoreView() instruction to save a snapshot
+ * of the current view and restore it when listeners are invoked. This allows
+ * walking the declaration view tree in listeners to get vars from parent views.
+ */
+export function getCurrentView(): LViewData {
```
Comment (JoostK) [inline packages/core/src/render3/instructions.ts]: `LViewData` is currently not exported, not even as internal symbol.

File: packages/core/src/render3/instructions.ts [hunk 2]
```diff
@@ -108,11 +108,31 @@ export function getCurrentSanitizer(): Sanitizer|null {
...
+/**
+ * Restores `contextViewData` to the given LViewData instance.
+ *
+ * Used in conjunction with the getCurrentView() instruction to save a snapshot
+ * of the current view and restore it when listeners are invoked. This allows
+ * walking the declaration view tree in listeners to get vars from parent views.
+ *
+ * @param viewToRestore The LViewData instance to restore.
+ */
+export function restoreView(viewToRestore: LViewData) {
```
Comment (JoostK) [inline packages/core/src/render3/instructions.ts]: Same here ;)

**Review by JoostK — COMMENTED** [review]
Why not let the `listener` instruction attach the current view to the function and restore it when dispatching the listener?

The callback is already wrapped in `wrapListenerWithDirtyAndDefault` and `wrapListenerWithDirtyLogic` which has access to the `LViewData` at that moment in time, so just restoring it there seems straightforward. Or am I missing something? 😊

**Comment by kara** [comment]
@Joostk We are getting rid of that wrapper function soon, so that is not a viable solution.

File: packages/core/src/render3/instructions.ts (hunk 1 above)
Comment (mhevery) [reply packages/core/src/render3/instructions.ts]: Add brand interface, change comment

**Review by mhevery — APPROVED** [review]

**Comment by kara** [comment]
[presubmit](https://test.corp.google.com/ui#id=OCL:206838738:BASE:206839034:1533076253049:b3829062)

**Pull request closed without merging**